### PR TITLE
avr: use new cross compilation infrastructure

### PR DESCRIPTION
**Pull request by matthewbauer**
Gets rid of:
-  avrbinutils
-  avrgcc

to replace with:
-  pkgsCross.avr.buildPackages.binutils
-  pkgsCross.avr.buildPackages.gcc

### CODE REVIEW

**Comment by dtzWill** [comment]
:+1: :heart:

**Review by Ericson2314 — APPROVED** [review]
Gorgeous!

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build pkgsCross.avr.hello

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build pkgsCross.arduino-avr.hello

I need to figure out a way that we don't break qmk_firmware, which looks like it uses Nix:

https://github.com/qmk/qmk_firmware

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build pkgsCross.arduino-uno.hello

**Comment by Ericson2314** [comment]
Oh I recall our GCC derivation doing something funny with limits.h

**Comment by matthewbauer** [comment]
@ericson2314 It looks like it's coming from this line:

https://github.com/nixos/nixpkgs/blob/master/pkgs/development/compilers/gcc/builder.sh#L142

We've always assumed we had limits.h, but that's not always the case. Maybe remove that line from the builder? (later on of course, trying to avoid mass rebuilds here)

**Comment by matthewbauer** [comment]
That is we overwrite the builtin LIMITS_H_TEST with that flag, causing it to always be true:

https://github.com/gcc-mirror/gcc/blob/master/gcc/Makefile.in#L535

But it looks like we can't avoid the mass rebuild unfortunately!

**Comment by matthewbauer** [comment]
@GrahamcOfBorg build pkgsCross.arduino-uno.hello

**Comment by matthewbauer** [comment]
Pinging a few people who have used avrgcc to see what can be done to make this transition easier:

Specifically how can we get multiple compilers to work with cc-wrapper? Without knowing ICU and MCU ahead of time we are kind of stuck...

- @mguentner
- @acowley 
- @gesturbjarkason

**Comment by matthewbauer** [comment]
Just squashed commits and cleaned things up. Interested in getting any feedback on this!

File: pkgs/applications/audio/axoloti/default.nix [hunk 1]
```diff
@@ -20,7 +21,7 @@ stdenv.mkDerivation rec {
     sha256 = "0lb5s8pkj80mqhsy47mmq0lqk34s2a2m3xagzihalvabwd0frhlj";
   };
 
-  buildInputs = [ makeWrapper unzip gcc-arm-embedded dfu-util-axoloti jdk ant libfaketime ];
+  buildInputs = [ makeWrapper unzip gcc-arm-embedded binutils-arm-embedded dfu-util-axoloti jdk ant libfaketime ];
```
Comment (Ericson2314) [inline pkgs/applications/audio/axoloti/default.nix]: Some of these should be nativeBuildInputs but whatever, one problem at a time :)

**Review by Ericson2314 — APPROVED** [review]

**Comment by danielfullmer** [comment]
Very cool. Thanks for working on this. I can at least confirm this works with my own custom keyboard firmware based on `qmk_firmware`. If interested, see: https://github.com/danielfullmer/nixos-config/commit/27447be5cc69c30b44e34ec379130183a3d16a5d

**Comment by matthewbauer** [comment]
Merging this now. Apologies if this breaks anything. I will be available to fix any potential breakages this weekend. Please report any issues as the appear!

**Comment by chloekek** [comment]
It was difficult for me to figure out how to find avr-gcc after this change, since command-not-found no longer reports which package provides it. Is there anything we can do about that?

**Pull request merged**